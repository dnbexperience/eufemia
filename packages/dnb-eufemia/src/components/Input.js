/**
 * ATTENTION: This file is auto generated by using "prepareTemplates".
 * Do not change the content!
 *
 */

/**
 * Library Index input to autogenerate all the components and patterns
 * Used by "prepareInputs"
 */

import Input from './input/Input'
export * from './input/Input'
export default Input
