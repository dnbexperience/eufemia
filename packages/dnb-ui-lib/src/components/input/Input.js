/**
 * Web Input Component
 *
 */

import React, { PureComponent } from 'react'
import PropTypes from 'prop-types'
import classnames from 'classnames'
import { DefaultIconSize } from '../icon'
import Button from '../button/Button'
import FormLabel from '../form-label/FormLabel'
import FormStatus from '../form-status/FormStatus'
import {
  registerElement,
  validateDOMAttributes,
  processChildren,
  pickRenderProps,
  dispatchCustomElementEvent
} from '../../shared/component-helper'

const renderProps = {
  on_change: null,
  on_submit: null,
  on_focus: null,
  on_blur: null,
  on_submit_focus: null,
  on_submit_blur: null
}

export const propTypes = {
  type: PropTypes.string,
  size: PropTypes.string,
  value: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
  id: PropTypes.string,
  label: PropTypes.string,
  status: PropTypes.string,
  status_state: PropTypes.string,
  autocomplete: PropTypes.oneOf(['on', 'off']),
  search_button_title: PropTypes.string,
  placeholder: PropTypes.string,
  description: PropTypes.string,
  align: PropTypes.string,
  extra_information: PropTypes.string,
  disabled: PropTypes.oneOfType([PropTypes.string, PropTypes.bool]),
  class: PropTypes.string,
  input_class: PropTypes.string,
  input_typo: PropTypes.string,
  attributes: PropTypes.oneOfType([PropTypes.string, PropTypes.object]),

  // React props
  className: PropTypes.string,
  inputElement: PropTypes.func,
  children: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.node,
    PropTypes.func
  ]),

  // Web Component props
  custom_element: PropTypes.object,
  custom_method: PropTypes.func,
  on_change: PropTypes.func,
  on_submit: PropTypes.func,
  on_focus: PropTypes.func,
  on_blur: PropTypes.func,
  on_submit_focus: PropTypes.func,
  on_submit_blur: PropTypes.func
}

export const defaultProps = {
  type: 'text',
  size: null,
  value: null,
  id: null,
  label: null,
  status: null,
  status_state: 'error',
  autocomplete: 'off',
  search_button_title: '',
  placeholder: null,
  description: null,
  align: null,
  extra_information: null,
  disabled: false,
  input_class: null,
  input_typo: 'dnb-typo-number--lining',
  class: null,
  attributes: null,

  // React props
  className: null,
  inputElement: null,
  children: null,

  // Web Component props
  custom_element: null,
  custom_method: null,
  ...renderProps
}

/**
 * The input component is an umbrella component for all inputs which share the same style as the classic `text` input field. Radio buttons and other form elements are not included here.
 */
export default class Input extends PureComponent {
  static tagName = 'dnb-input'
  static propTypes = propTypes
  static defaultProps = defaultProps
  static renderProps = renderProps

  static enableWebComponent() {
    registerElement(Input.tagName, Input, defaultProps)
  }

  static getDerivedStateFromProps(props, state) {
    const value = Input.getValue(props)
    if (
      state._listenForPropChanges &&
      value !== null &&
      value !== state.value
    ) {
      state.value = value
    }
    if (props.disabled) {
      state.inputState = 'disabled'
    }
    state._listenForPropChanges = true
    return state
  }

  static getValue(props) {
    if (props.value) return props.value
    return processChildren(props)
  }

  state = { inputState: 'virgin', value: null }

  constructor(props) {
    super(props)

    this._ref = React.createRef()
    this._id = props.id || `dnb-input-${Math.round(Math.random() * 999)}` // cause we need an id anyway

    // pass along all props we wish to have as params
    this.renderProps = pickRenderProps(props, renderProps)

    // make sure we dont trigger getDerivedStateFromProps on startup
    this.state._listenForPropChanges = true
    this.state.value = Input.getValue(props)

    // This is only to show, that we also can add a custom method to the custom element, and call it like so:
    // this._ref.current.focus() // current would be the Custom Element - we may delay it for 1 tick, to make sure we call if after it is set
    // We also have to ahve this method in here
    // setCustomElementMethod(this, 'focus', () => {
    //   setTimeout(() => {
    //     try {
    //       this._ref.current.focus()
    //     } catch (e) {
    //       //
    //     }
    //   }, 100)
    // })
  }
  onFocusHandler = event => {
    const value = event.target.value
    this.setState({
      value,
      _listenForPropChanges: false,
      inputState: 'focus'
    })
    dispatchCustomElementEvent(this, 'on_focus', { value, event })
  }
  onBlurHandler = event => {
    const value = event.target.value
    this.setState({
      value,
      _listenForPropChanges: false,
      inputState: 'dirty'
    })
    dispatchCustomElementEvent(this, 'on_blur', { value, event })
  }
  onChangeHandler = event => {
    const value = event.target.value
    this.setState({ value, _listenForPropChanges: false })
    event.preventDefault()
    dispatchCustomElementEvent(this, 'on_change', { value, event })
  }
  onKeyDownHandler = event => {
    if (event.key === 'Enter') {
      const value = event.target.value
      event.preventDefault()
      dispatchCustomElementEvent(this, 'on_submit', { value, event })
    }
  }
  render() {
    const {
      type,
      size,
      label,
      status,
      status_state,
      disabled,
      placeholder,
      description,
      align,
      input_class,
      input_typo,
      extra_information
    } = this.props

    let { autocomplete } = this.props
    if (type === 'search') {
      autocomplete = null
    }
    const id = this._id

    const classes = classnames(
      'dnb-input',
      `dnb-input--${type}`, //type_modifier
      extra_information ? 'dnb-input--has-extra-information' : null,
      size ? 'dnb-input--' + size : '',
      type === 'search' ? 'dnb-input__input--search' : null,
      align ? `dnb-input__align--${align}` : null,
      status ? `dnb-input__status--${status_state}` : null,
      this.props.class,
      this.props.className
    )

    const { inputElement: Elem, ...renderProps } = this.renderProps

    const inputParams = {
      ...renderProps,
      className: classnames('dnb-input__input', input_typo, input_class),
      autoComplete: autocomplete,
      value: this.state.value || '',
      type,
      id,
      // align,
      disabled,
      name: id,
      onChange: this.onChangeHandler,
      onKeyDown: this.onKeyDownHandler,
      onFocus: this.onFocusHandler,
      onBlur: this.onBlurHandler
    }

    // also used for code markup simulation
    validateDOMAttributes(this.props, inputParams)

    if (description) {
      inputParams['aria-describedby'] = id + '-description'
    }

    const shellParams = {
      'data-input-state': this.state.inputState,
      'data-has-content':
        String(this.state.value || '').length > 0 ? 'true' : 'false'
    }

    return (
      <span className={classes}>
        {label && (
          <FormLabel for_id={id} text={label} disabled={disabled} />
        )}

        <span className="dnb-input__shell" {...shellParams}>
          {(type === 'text' || type === 'number' || type === 'search') &&
            ((typeof Elem === 'function' ? (
              <Elem innerRef={this._ref} {...inputParams} />
            ) : null) || <input ref={this._ref} {...inputParams} />)}

          {type === 'search' && (
            <Submit
              {...this.props}
              value={inputParams.value}
              title={this.props.search_button_title}
            />
          )}

          {placeholder && (
            <span
              className={classnames(
                'dnb-input__placeholder',
                align ? `dnb-input__align--${align}` : null
              )}
            >
              {placeholder}
            </span>
          )}
        </span>

        {this.props.description && (
          <span
            className="dnb-input__description"
            id={id + '-description'}
          >
            {this.props.description}
          </span>
        )}

        {extra_information && (
          <span className="dnb-input__extra-information">
            {extra_information}
          </span>
        )}

        {status && status !== 'true' && (
          <FormStatus for_id={id} text={status} status={status_state} />
        )}
      </span>
    )
  }
}

class Submit extends PureComponent {
  static propTypes = {
    title: PropTypes.string.isRequired,
    value: PropTypes.string.isRequired,
    icon_size: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),

    // React props
    onSubmit: PropTypes.func,
    onSubmitFocus: PropTypes.func,
    onSubmitBlur: PropTypes.func,

    // Web Component props
    on_submit: PropTypes.func,
    on_submit_focus: PropTypes.func,
    on_submit_blur: PropTypes.func
  }

  static defaultProps = {
    title: null,
    icon_size: DefaultIconSize,

    // React props
    onSubmit: null,
    onSubmitFocus: null,
    onSubmitBlur: null,

    // Web Component props
    on_submit: null,
    on_submit_focus: null,
    on_submit_blur: null
  }

  state = { focusState: 'virgin' }

  onFocusHandler = event => {
    const value = this.props.value
    this.setState({
      focusState: 'focus'
    })
    dispatchCustomElementEvent(this, 'on_submit_focus', { value, event })
  }
  onBlurHandler = event => {
    const value = this.props.value
    this.setState({
      focusState: 'dirty'
    })
    dispatchCustomElementEvent(this, 'on_submit_blur', { value, event })
  }
  onSubmitHandler = event => {
    const value = this.props.value
    dispatchCustomElementEvent(this, 'on_submit', { value, event })
  }
  render() {
    const { title } = this.props
    const params = {
      type: 'submit',
      title
    }

    // also used for code markup simulation
    validateDOMAttributes(this.props, params)

    return (
      <span
        className="dnb-input__search-submit"
        data-input-state={this.state.focusState}
      >
        <Button
          className="dnb-input__search-submit__button"
          variant="secondary"
          onClick={this.onSubmitHandler}
          onFocus={this.onFocusHandler}
          onBlur={this.onBlurHandler}
          icon="search"
          {...params}
        />
      </span>
    )
  }
}
