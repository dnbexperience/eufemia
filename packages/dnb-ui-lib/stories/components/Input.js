/**
 * dnb-ui-lib Component Story
 *
 */

import React, { useState, useEffect } from 'react'
import { Wrapper, Box } from '../helpers'
import styled from '@emotion/styled'

import {
  Input,
  InputMasked,
  Button,
  FormSet,
  FormRow,
  FormLabel
} from '../../src/components'

const CustomStyle = styled.div`
  p {
    background-color: rgba(213, 30, 149, 0.25);
  }
`

export default [
  'Inputs',
  () => (
    <CustomStyle>
      <Wrapper>
        <Box>
          <Input
            label="Choose file:"
            type="file"
            accept="image/png, image/jpeg"
          />
        </Box>
        <Box>
          <CustomInput />
        </Box>
        <Box>
          <FormSet>
            <FormRow
              indent
              indent_offset="x-large"
              wrap
              label="Long label labwl Adipiscing mauris dis proin nec Condimentum egestas class blandit netus non a suscipit id urna:"
            >
              <Input label="Input A:" top="small" right="small" />
              <Input label="Input B:" top="small" right="small" />
              <Input label="Input C:" top="small" right="small" />
            </FormRow>
          </FormSet>
        </Box>
        <Box>
          <FormRow
            label="Vertical label:"
            label_direction="vertical"
            // vertical
          >
            <Input label="Input label A:" right="small" />
            <Input label="Input label B:" />
          </FormRow>
        </Box>
        <Box>
          <FormSet direction="vertical">
            <FormRow label="Legend:">
              <Input label="Vertical 1:" />
              <Input label="Vertical 2:" stretch top="small" />
            </FormRow>
          </FormSet>
        </Box>
        <Box>
          <FormSet vertical>
            <FormRow label="Legend:">
              <Input label="Vertical 1:" />
              <Input label="Vertical 2:" stretch top="small" />
            </FormRow>
          </FormSet>
        </Box>
        <Box>
          <Input
            label="Vertical label:"
            value="Stretch me ..."
            stretch
            label_direction="vertical"
          />
        </Box>
        <Box>
          Text
          <Input
            selectall
            label="Label:"
            on_change={event => {
              console.log('on_change', event)
            }}
          >
            Input ...
          </Input>
          Text
        </Box>
        <Box>
          Text
          <FormLabel>FormLabel:</FormLabel>
          <Input>Input ...</Input>
          Text
        </Box>
        <Box>
          <p className="dnb-p">
            <Input
              label="ReadOnly:"
              placeholder="Placeholder ..."
              readOnly
            />
          </p>
        </Box>
        <Box>
          <Input
            label="Search:"
            type="search"
            align="right"
            stretch
            submit_button_title="Search"
            placeholder="Search text placeholder"
          />
          <Input
            label="Search:"
            size="medium"
            type="search"
            align="right"
            stretch
            placeholder="Medium input with right aligned text"
          />
          <Input
            label="Search:"
            size="large"
            type="search"
            align="right"
            stretch
            placeholder="Large input with right aligned text"
          />
        </Box>
        <Box>
          <Input
            disabled
            status="Error"
            label="Disabled search:"
            type="search"
            submit_button_title="Search"
            placeholder="Search text placeholder"
          />
        </Box>
        <Box>
          <Input
            label="Input with status:"
            status="Message to the user"
            value="Input value with status"
          />
        </Box>
        <Box>
          <Input
            label="Input with description:"
            description="Description to the user"
            value="Input value with status"
          />
        </Box>
        <Box>
          <InputMasked
            label="Masked:"
            autocomplete="off"
            // value="1000000"
            mask={[
              '+',
              /[4]/, // have to start with 4
              /[5-7]/, // can be 5,6 or 7
              ' ',
              '/',
              ' ',
              /[49]/, // have to start with 4 or 9
              /\d/,
              /\d/,
              ' ',
              /\d/,
              /\d/,
              ' ',
              /\d/,
              /\d/,
              /\d/
            ]}
            show_mask="true"
          />
        </Box>
        <Box>
          <form
            onSubmit={event => {
              // console.log('onSubmit', event)
              event.preventDefault()
              // event.persist()
            }}
          >
            <InputUpdate />
            <Button
              text="Submit"
              type="submit"
              on_click={event => {
                console.log('on_click', event)
              }}
              onClick={event => {
                console.log('onClick', event)
              }}
            />
          </form>
        </Box>
      </Wrapper>
    </CustomStyle>
  )
]

const InputUpdate = () => {
  const [initValue, setNewValue] = useState('Input ...')

  useEffect(() => {
    const timeoutId = setTimeout(() => {
      setNewValue('')
    }, 1e3)
    return () => clearTimeout(timeoutId)
  })

  return (
    <Input
      label="Label:"
      on_change={({ value }) => {
        console.log('on_change', value)
      }}
      onChange={({ value }) => {
        console.log('onChange', value)
      }}
      on_submit={({ value }) => {
        console.log('on_submit', value)
      }}
      onSubmit={({ value }) => {
        console.log('onSubmit', value)
      }}
      value={initValue}
    />
  )
}

const CustomInput = () => {
  const [value, setValue] = useState('2019-02-15')
  return (
    <>
      <Input
        value={value}
        on_change={({ value }) => {
          console.log('on_change', value)
          setValue(value)
        }}
        on_state_update={({ value }) => {
          console.warn('on_state_update', value)
          setValue(value)
        }}
        right
      />
      <Button
        text="Reset"
        on_click={() => {
          setValue('123')
        }}
      />
    </>
  )
}
