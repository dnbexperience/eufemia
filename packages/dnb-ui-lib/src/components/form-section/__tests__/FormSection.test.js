/**
 * Component Test
 *
 */

import React from 'react'
import {
  mount,
  fakeProps,
  axeComponent,
  toJson,
  loadScss
} from '../../../core/jest/jestSetup'
import Component from '../FormSection'
import Input from '../../input/Input'

// just to make sure we re-run the test in watch mode due to changes in theese files
import _form_section from '../style/_form-section.scss' // eslint-disable-line
import dnb_form_section from '../style/dnb-form-section.scss' // eslint-disable-line
import dnb_form_section_theme_ui from '../style/themes/dnb-form-section-theme-ui.scss' // eslint-disable-line

const props = fakeProps(require.resolve('../FormSection'), {
  optional: true
})
props.direction = 'horizontal'

describe('FormSection component', () => {
  const Comp = mount(<Component {...props} />)

  it('have to match snapshot', () => {
    expect(toJson(Comp)).toMatchSnapshot()
  })

  it('should forward unlisted attributes like "aria-hidden"', () => {
    const Comp = mount(<Component {...props} for_id="input" aria-hidden />)
    expect(Comp.find('section[aria-hidden]').exists()).toBe(true)
    expect(
      Comp.find('section[aria-hidden]')
        .instance()
        .getAttribute('aria-hidden')
    ).toBe('true')
  })

  it('should validate with ARIA rules', async () => {
    expect(await axeComponent(Comp)).toHaveNoViolations()
  })

  it.skip('should validate with ARIA rules as a section with a input', async () => {
    const SectionComp = mount(<Component {...props} for_id="input" />)
    const InputComp = mount(<Input id="input" value="some value" />)
    expect(await axeComponent(SectionComp, InputComp)).toHaveNoViolations()
  })
})

describe('FormSection scss', () => {
  it('have to match snapshot', () => {
    const scss = loadScss(
      require.resolve('../style/dnb-form-section.scss')
    )
    expect(scss).toMatchSnapshot()
  })
  it('have to match default theme snapshot', () => {
    const scss = loadScss(
      require.resolve('../style/themes/dnb-form-section-theme-ui.scss')
    )
    expect(scss).toMatchSnapshot()
  })
})
