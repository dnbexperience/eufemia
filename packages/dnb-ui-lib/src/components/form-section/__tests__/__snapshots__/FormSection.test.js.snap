// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FormSection component have to match snapshot 1`] = `
<FormSection
  class="class"
  className="className"
  render_content={[Function]}
  spacing="spacing"
  style="style"
>
  <section
    className="dnb-form-section dnb-form-section--style dnb-form-section--spacing-spacing className class"
  >
    children
  </section>
</FormSection>
`;

exports[`FormSection scss have to match default theme snapshot 1`] = `
"/*
 * FormSection theme
 *
 */
.dnb-form-section--divider::after {
  color: var(--color-white);
  box-shadow: 100vw 0 0 0 currentColor, 0 0.0625rem 0 0 var(--color-mint-green-50), 100vw 0.0625rem 0 0 var(--color-mint-green-50), 0 -0.0625rem 0 0 var(--color-mint-green-50), 100vw -0.0625rem 0 0 var(--color-mint-green-50); }

.dnb-form-section--mint-green-12::after {
  color: var(--color-mint-green-12); }

.dnb-form-section--white::after {
  color: var(--color-white); }

.dnb-form-section--mint-green::after {
  color: var(--color-mint-green); }

.dnb-form-section--emerald-green {
  color: var(--color-mint-green); }
  .dnb-form-section--emerald-green::after {
    color: var(--color-emerald-green); }
  .dnb-form-section--emerald-green .dnb-code {
    color: var(--color-emerald-green); }

.dnb-form-section--signal-orange {
  color: var(--color-ocean-green); }
  .dnb-form-section--signal-orange::after {
    color: var(--color-signal-orange); }
"
`;

exports[`FormSection scss have to match snapshot 1`] = `
"/*
* DNB FormSection
*
*/
/**
 * This file is only used to make components independent
 * so that they can get imported individually, without the core styles
 *
 */
/*
 * Utilities
 */
/*
 * Scopes
 *
 */
/*
 * Document Reset
 *
 */
.dnb-form-section {
  font-family: var(--font-family-default);
  font-weight: var(--font-weight-default);
  font-size: 1rem;
  font-style: normal;
  line-height: 1.5rem;
  color: var(--color-black-80, #333);
  -webkit-tap-highlight-color: rgba(0, 0, 0, 0);
  font-variant-numeric: lining-nums;
  font-feature-settings: \\"lnum\\";
  /**
  * 1. Remove repeating backgrounds in all browsers (opinionated).
  * 2. Add border box sizing in all browsers (opinionated).
  */
  /**
  * 1. Add text decoration inheritance in all browsers (opinionated).
  * 2. Add vertical alignment inheritance in all browsers (opinionated).
  */
  margin: 0;
  padding: 0; }
  .dnb-form-section *,
  .dnb-form-section ::before,
  .dnb-form-section ::after {
    background-repeat: no-repeat;
    /* 1 */
    box-sizing: border-box;
    /* 2 */ }
  .dnb-form-section ::before,
  .dnb-form-section ::after {
    text-decoration: inherit;
    /* 1 */
    vertical-align: inherit;
    /* 2 */ }

/*
 * FormSection component
 *
 */
.dnb-form-section {
  position: relative; }
  .dnb-form-section::after {
    content: '';
    position: absolute;
    z-index: -1;
    left: -50vw;
    top: 0;
    width: 100vw;
    height: 100%;
    color: var(--color-mint-green-12);
    background-color: currentColor;
    box-shadow: 100vw 0 0 0 currentColor; }
  .dnb-form-section--spacing,
  .dnb-spacing .dnb-form-section--spacing {
    padding-top: 2rem;
    padding-bottom: 2rem; }
  .dnb-form-section--spacing-small,
  .dnb-spacing .dnb-form-section--spacing-small {
    padding-top: 1rem;
    padding-bottom: 1rem; }
  .dnb-form-section--spacing-large,
  .dnb-spacing .dnb-form-section--spacing-large {
    padding-top: 3rem;
    padding-bottom: 3.5rem; }
"
`;
